blockwork/workflows/workflow.py

```diff
@@ -134,35 +140,94 @@ def gather(self, config: Config) -> Iterable[tuple[Config, list[Transform], list
             for transform in transforms:
                 dependency_map[transform] = set()
+                dependent_map[transform] = set()
                 for interface in transform.real_output_interfaces:
                     output_interfaces.append(interface)
-            targets += target_transforms
+            targets.update(target_transforms)
             
         # We only need to look at output interfaces since an output must
         # exist in order for a dependency to exist
         for interface in output_interfaces:
             input_transform = cast(Transform, interface.input_transform)
             for output_transform in interface.output_transforms:
                 dependency_map[output_transform].add(input_transform)
+                dependent_map[input_transform].add(output_transform)
+        
+        return targets, dependency_map, dependent_map
+    
+    def _run(self,
+             ctx: Context,
+             targets: set[Transform],
+             dependency_map: dict[Transform, set[Transform]],
+             dependent_map: dict[Transform, set[Transform]]):
+        '''
+        Run the workflow from transform dependency data.
+        '''
 
-        # Run everything in order serially
-        scheduler = Scheduler(dependency_map, targets=targets)
-        while scheduler.incomplete:
-            for transform in scheduler.schedulable:
-                scheduler.schedule(transform)
-                # Note this message is info for now for demonstrative purposes only
-                logging.info("Running transform: %s", transform)
-                transform.run(ctx)
-                scheduler.finish(transform)
+        # Record the transforms we pulled from the cache
+        fetched_transforms: set[Transform] = set()
+        # Record the transforms that don't need to be run since they were were
+        # pulled from the cahce or only transforms pulled from the cache 
+        # depend on them.
+        skipped_transforms: set[Transform] = set()
+        # Record the transforms we actually ran
+        run_transforms: set[Transform] = set()
+        # And those that made it into the cache
+        stored_transforms: set[Transform] = set()
+
+        # Run in reverse order, pulling from the cache if items exits
+        cache_scheduler = Scheduler(dependency_map, targets=targets, reverse=True)
+        while cache_scheduler.incomplete:
+            for transform in cache_scheduler.schedulable:
+                cache_scheduler.schedule(transform)
+                if transform not in targets:
+                    if not (dependent_map[transform] - skipped_transforms):
+                        skipped_transforms.add(transform)
+                    elif Cache.fetch_transform(ctx, transform):
+                        logging.info("Fetched transform from cache: %s", transform)
+                        fetched_transforms.add(transform)
+                        skipped_transforms.add(transform)
```

Comment:
Given that accumulating `skipped_transforms` mostly seems to assist debug/testing, is there much point appending to both skipped and fetched in this `elif` as the rest of the code excludes fetched from skipped anyway?